rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,429 @@
+- (+) Takes up less vertical space
+- (-) Visually confusing when every line starts with a `#` character.
+    - (-) Potential visual conflicts with the content that is markdown (e.g. with headings).
+- (+) Indentation of the content is clear.
+
+</details>
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Changes the existing comments inside the code base
+
+This could be automated.
+
+Also, the migration could be performed piecemal, starting perhaps with `nixpkgs.lib`.
+
+See future work.
+
+## Tooling to produce the nixpkgs manual
+
+Change or write a new tool used to produce Nixpkgs library function documentation. Because this is a breaking change.
+
+See future work.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## All considered outer formats
+
+| Property / Approach | `##` | `/** */` | `Javadoc` | `/*\|` or `/*^`  | All comments are doc-comments |
+|---|---|---|---|---|---|
+| Inspired by | Rust | Current nixpkgs.lib | C++/Java/Javascript | Haskell Haddock | Current nixpkgs.lib |
+| Changes the existing code by | Much | Less | Even More | Less | ? |
+| Needs Termination | No | Yes | Yes | Yes | ? |
+| Indentation | Clear | like Nix's multiline strings, thus **Intuitive** | Clear | ? | ? |
+| Needs vertical space  | No | Yes | Yes | Yes | ? |
+| Visual distinction from comments | High | Low | Medium | Medium | No distinction at all |
+| Needs Autocompletion (Language Server) to continue the next line. | Yes | No | Yes | No | ? |
+| Punctuation Variations / Amount of different special characters | 1 (Less) | 2 (Medium) | 2 (Medium) | 3 (More) | None |
+| Markdown compatibility (also depends on indentation clarity) | **Visual conflicts with headings `# Title`** | Good | Medium | Good | Only with multiline comments |
+| breaks when interrupted with newlines | Yes | No | ? | No | |
+| Simplicity (Brainload) | Medium | Simple | Complex | More Complex | |
+
+### Refactoring note
+
+**Observing**: From a refactoring perspective, it might also be interesting to see how many conflicts the different formats would cause.
+
+nixpkgs comments:
+
+- `##` ~4k usages
+- `#` ~20k usages
+- `/*` ~6k usages
+- `/**` 160 usages 
+
+Choosing `/**` or subsets would cause minor conflicts within current nixpkgs. While this is NOT the main reason for the final decision, it MUST be considered.
+
+## Just free text as a content format
+
+While this allows the most freedom, it is usually considered the best option, not creating any restrictions.
+
+- [RFC72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md) defines commonMark as the official documentation format.
+This is why we decided to follow this convention. Writing plain text is still possible.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Migration path for nixpkgs comments.
+- How to document the `arguments`. Should there be some markdown equivalent to `@param` i.e. `# Params`?
+    - Possible answer: Future RFC might need to clarify that.
```

Comment:
Formally the RFC should make *some* decision, where postponing is a valid decision.

However, in my opinion that should block the design of the builtins. The language is meant to be stable. If we accept an incomplete design, we'll have to support an incomplete API **in addition to** a complete one, because the language is supposed to be reproducible - very backwards compatible. I can see an argument about incremental improvements, but I'm really not a fan of adding more tech debt to Nix.

I would suggest to only postpone this question if you have a very good reason to do so.

(nit: remove backticks around `arguments`)

Reply:
So in your opinion we should specify all reserved keywords such as e.g., `arguments`? 

The documentation team disliked the idea of having reserved keywords and proposed of leaving this open to the user. Nix will not be responsible for rendering anything in my opinion. The only role nix has there is to ensure that the user can retrieve the content. Rendering the comment itself can be done with any markdown tool. 

I had a talk with @tomberek and we both think that adding this as a `builtin` to the language is not the correct way. Instead we should maybe add an additional (yet experimental) command which can be called with something like: `nix doc --file my.nix path.to.foo`. Adding the builtin would (A) change the language and (B) allow nix to introspect itself which could be dangerous.

Reply:
> So in your opinion we should specify all reserved keywords such as e.g., `arguments`?

Wait, does this refer to the nit?

> The documentation team disliked the idea of having reserved keywords and proposed of leaving this open to the user.

Fine by me.

-----

I was also thinking of syntax such as the following, which would affect a documentation builtin.

```nix
/** A function for using a thing. It works in a certain way. */
{ /** the thing */
  it,
  /** arbitrary other things to make it work differently */
  ...
}:
```

-----



> Adding the builtin would (A) change the language

That was also my concern. Adding to the language is fine, but it's hard-to-impossible to change anything after.

> and (B) allow nix to introspect itself which could be dangerous.

We already have introspection, so we've already accepted a degree of this danger.

> experimental feature

This allows the design to change and stabilize over time. If we're too concerned about introspection, we could perhaps turn it into a non-experimental setting that's off by default.

Reply:
What do you think about dropping the builtin idea and having a cli which should be easier to change in the future?